lib/Sema/TypeCheckDecl.cpp

```diff
@@ -1957,8 +1957,17 @@ void TypeChecker::validateDecl(OperatorDecl *OD) {
 }
 
 bool swift::doesContextHaveValueSemantics(DeclContext *dc) {
-  if (Type contextTy = dc->getDeclaredInterfaceType())
-    return !contextTy->hasReferenceSemantics();
+  if (Type contextTy = dc->getDeclaredInterfaceType()) {
+    // If the decl context is an extension, then it could be imposing a class
+    // constraint (ex: where Self: SomeClass). Make sure we include that
+    // in our check as well.
+    auto extensionRequiresClass = false;
+    if (auto ED = dyn_cast<ExtensionDecl>(dc)) {
+      extensionRequiresClass =
+          ED->getGenericSignature()->requiresClass(ED->getSelfInterfaceType());
+    }
+    return !contextTy->hasReferenceSemantics() && !extensionRequiresClass;
```

Comment:
I'm just noticing this now, sorry, but this will do all the work even if the immediate context type has reference semantics and it wasn't important to check the extension. Please use early returns.

Reply:
I think we should also be using `getSelfTypeInContext()` here as well, right? Then we won’t need the extension check!